lib/util/source-code.js

```diff
@@ -111,9 +111,12 @@ class SourceCode extends TokenStore {
      * Represents parsed source code.
      * @param {string} text - The source code text.
      * @param {ASTNode} ast - The Program node of the AST representing the code. This AST should be created from the text that BOM was stripped.
+     * @param {Object|null} parserServices - The parser srevices.
+     * @param {ScopeManager|null} scopeManager - The scope of this source code.
+     * @param {Object|null} visitorKeys - The visitor keys to traverse AST.
      * @constructor
      */
-    constructor(text, ast) {
+    constructor(text, ast, parserServices, scopeManager, visitorKeys) {
```

Comment:
One object parameter sounds fine to me. I just don't want to have too many positional parameters, because it could end up confusing to use (like the old `context.report` API).